File: C++/034. Find First and Last Position of Element in Sorted Array.cpp
class Solution {
private:
    int searchLeftBound(vector<int>& nums, int left, int right, int target)
    {
        int center;
        while(left<=right)
        {
            if(left==right){return left;}
            center =(left+right)/2;
            if(nums[center]<target)
            {
                left=center+1;
            }
            else
            {
                right=center;
            }
        }
        return left;
    }
    int searchRightBound(vector<int>& nums, int left, int right, int target)
    {
        int center;
        while(left<=right)
        {
            if(left==right){return left;}
            center =(left+right)/2;
            if(nums[center]>target)
            {
                right=center-1;
            }
            else
            {
                if (left!=center){left=center;}
                else
                {
                    return (nums[right]==target)? right:left;
                }
            }
        }
        return left;
    }

public:
    vector<int> searchRange(vector<int>& nums, int target) {
        vector<int> res(2,-1);
        int size = nums.size();
        int left=0, right= size-1, center, c1, c2;
        while(left<=right)
        {
            if(left==right)
            {
                if(nums[left]==target)
                {
                    res[0]=left;
                    res[1]=right;
                    return res;
                }
                return res;
            }
            center = (left+right)/2;
            if(nums[center]<target)
            {
                left = center+1;
            }
            else if(nums[center]>target)
            {
                right = center;
            }
            else{break;}
        }
        res[0] = searchLeftBound(nums, left, center, target);
        res[1] = searchRightBound(nums, center, right, target);
        return res;
    }
};

File: python/034. Find First and Last Position of Element in Sorted Array.py
class Solution(object):
    def searchRange(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[int]
        """
        n = len(nums)
        if n == 0: return [-1, -1]
        # find the first element nums[i] such taht nums[i] >= target
        left, right = 0, n-1
        while left <= right:
            mid = (left + right) / 2
            if nums[mid] >= target:
                right = mid - 1
            else:
                left = mid + 1
        if left == n or nums[left] != target: return [-1, -1]
        a = left
        # find the last element nums[i] such taht nums[i] <= target
        left, right = 0, n-1
        while left <= right:
            mid = (left + right) / 2
            if nums[mid] <= target:
                left = mid + 1
            else:
                right = mid - 1
        return [a, right]